yt_dlp/extractor/tumblr.py

```diff
@@ -107,116 +237,176 @@ class TumblrIE(InfoExtractor):
-            })
-
-        # Successful login
-        if '/dashboard' in urlh.geturl():
+        if not self._ACCESS_TOKEN:
+            self.report_warning('Cannot log in without access token')
             return
 
-        login_errors = self._parse_json(
-            self._search_regex(
-                r'RegistrationForm\.errors\s*=\s*(\[.+?\])\s*;', response,
-                'login errors', default='[]'),
-            None, fatal=False)
-        if login_errors:
-            raise ExtractorError(
-                'Unable to login: %s' % login_errors[0], expected=True)
-
-        self.report_warning('Login has probably failed')
+        try:
+            self._download_json(
+                self._OAUTH_URL, None, 'Logging in',
+                data=urlencode_postdata({
+                    'password': password,
+                    'grant_type': 'password',
+                    'username': username,
+                }), headers={
+                    'Content-Type': 'application/x-www-form-urlencoded',
+                    'Authorization': f'Bearer {self._ACCESS_TOKEN}',
+                })
+        except ExtractorError:
+            self.report_warning('Login failed')
 
     def _real_extract(self, url):
         m_url = self._match_valid_url(url)
         video_id = m_url.group('id')
         blog = m_url.group('blog_name')
 
-        url = 'http://%s.tumblr.com/post/%s/' % (blog, video_id)
+        url = f'http://{blog}.tumblr.com/post/{video_id}/'
         webpage, urlh = self._download_webpage_handle(url, video_id)
 
         redirect_url = urlh.geturl()
-        if 'tumblr.com/safe-mode' in redirect_url or redirect_url.startswith('/safe-mode'):
-            raise ExtractorError(
-                'This Tumblr may contain sensitive media. '
-                'Disable safe mode in your account settings '
-                'at https://www.tumblr.com/settings/account#safe_mode',
-                expected=True)
 
+        api_only = bool(self._search_regex(
+            r'(tumblr.com|^)/(safe-mode|login_required|blog/view)',
+            redirect_url, 'redirect', default=None))
+
+        if api_only and not self._ACCESS_TOKEN:
+            raise ExtractorError('Cannot get data for dashboard-only post without access token')
+
+        post_json = traverse_obj(
+            self._download_json(
+                f'https://www.tumblr.com/api/v2/blog/{blog}/posts/{video_id}/permalink',
+                video_id,
+                headers={
+                    'Authorization': f'Bearer {self._ACCESS_TOKEN}',
+                },
+                fatal=False) if self._ACCESS_TOKEN else None,
+            ('response', 'timeline', 'elements', 0), default={})
+        content_json = traverse_obj(
+            post_json, ('trail', 0, 'content'), ('content'), default=[])
+        video_json = next(
+            (item for item in content_json if item.get('type') == 'video'),
+            {})
+        media_json = video_json.get('media', {})
```

Comment:
when using .get for iterables, Don't pass the blank iter object as default.
rather do this 
`var = your_dict.get('key') or <iter blank object>`
for example here it should be `media_json = video_json.get('media') or {}`

Fix this at other places too

Reply:
Thanks. I wondered which was preferred while writing this, but it's not specified in CONTRIBUTING; perhaps it should be.

Reply:
this is done to handle `your_dict['key']` being literally `None`. It's applicable not just for iterators, but anywhere None or empty string can cause issues. While this specific case is not stated in contributing, similar situations are mentioned

---

Another thing to remember is not to try to iterate over `None`

Say you extracted a list of thumbnails into `thumbnail_data` using `try_get` and now want to iterate over them

```python
thumbnail_data = try_get(...)
thumbnails = [{
    'url': item['url']
} for item in thumbnail_data or []]  # correct
```

and not like:

```python
thumbnail_data = try_get(...)
thumbnails = [{
    'url': item['url']
} for item in thumbnail_data]  # incorrect
```

In the later case, `thumbnail_data` will be `None` if the field was not found and this will cause the loop `for item in thumbnail_data` to raise a fatal error. Using `for item in thumbnail_data or []` avoids this error and results in setting an empty list in `thumbnails` instead.

---


Assume that you have some another source you can extract `title` from, for example `og:title` HTML meta of a `webpage`. In this case you can provide a fallback scenario:

```python
title = meta.get('title') or self._og_search_title(webpage)
```

This code will try to extract from `meta` first and if it fails it will try extracting `og:title` from a `webpage`.

Reply:
I see, thank you.